src/utils/subgraph.ts

```diff
@@ -0,0 +1,120 @@
+import { getApolloClientInstance } from '../providers/ApolloClientProvider'
+import BigNumber from 'bignumber.js'
+
+const freQuery = gql`
+  query AssetFrePrice($datatoken_in: [String!]) {
+    fixedRateExchanges(orderBy: id, where: { datatoken_in: $datatoken_in }) {
+      rate
+      id
+      datatoken {
+        id
+        address
+      }
+    }
+  }
+`
+
+const poolQuery = gql`
+  query AssetPoolPrice($datatokenAddress_in: [String!]) {
+    pools(where: { datatokenAddress_in: $datatokenAddress_in }) {
+      spotPrice
+      id
+      datatokenAddress
+    }
+  }
+`
+
+const previousOrderQuery = gql`
+  query AssetPreviousOrder($id: String!, $account: String!) {
+    tokenOrders(
+      first: 1
+      where: { datatokenId: $id, payer: $account }
+      orderBy: timestamp
+      orderDirection: desc
+    ) {
+      timestamp
+      tx
+    }
+  }
+`
+
+async function fetchData(
+  query: DocumentNode,
+  variables: any
+): Promise<ApolloQueryResult<any>> {
+  try {
+    const client = getApolloClientInstance()
+    const response = await client.query({
+      query: query,
+      variables: variables
+    })
+    return response
+  } catch (error) {
+    console.error('Error fetchData: ', error.message)
+  }
+}
+
+export async function getPreviousOrders(
+  id: string,
+  account: string,
+  assetTimeout: string
+): Promise<string> {
+  const variables = {
+    id: id,
+    account: account
+  }
+  const fatchedPreviousOrders: any = await fetchData(
+    previousOrderQuery,
+    variables
+  )
+  if (fatchedPreviousOrders.data?.tokenOrders?.length === 0) return null
+  if (assetTimeout === 'Forever') {
+    return fatchedPreviousOrders?.data?.tokenOrders[0]?.tx
+  } else {
+    const expiry = new BigNumber(
+      fatchedPreviousOrders?.data?.tokenOrders[0]?.timestamp
+    ).plus(assetTimeout)
+    const unixTime = new BigNumber(Math.floor(Date.now() / 1000))
+    if (unixTime.isLessThan(expiry)) {
+      return fatchedPreviousOrders?.data?.tokenOrders[0]?.tx
+    } else {
+      return null
+    }
+  }
+}
+
+export async function getAssetPrice(assets: DDO[]): Promise<any> {
+  const priceList: any = {}
+  const poolPriceAssets: string[] = []
+  const poolDTadressDID: any = {}
+  const frePriceAssets: string[] = []
+  const freDTadressDID: any = {}
+  for (const ddo of assets) {
+    if (ddo.price?.type === 'pool') {
+      poolDTadressDID[ddo?.dataToken.toLowerCase()] = ddo.id
+      poolPriceAssets.push(ddo?.dataToken.toLowerCase())
+    } else if (ddo.price?.type === 'exchange') {
+      freDTadressDID[ddo?.dataToken.toLowerCase()] = ddo.id
+      frePriceAssets.push(ddo?.dataToken.toLowerCase())
+    } else {
+      priceList[ddo.id] = 'none'
+    }
+  }
+  const freVariables = {
+    datatoken_in: frePriceAssets
+  }
+  const poolVariables = {
+    datatokenAddress_in: poolPriceAssets
+  }
+  const poolPriceResponse: any = await fetchData(poolQuery, poolVariables)
+  for (const poolPirce of poolPriceResponse.data?.pools) {
```

Comment:
small typo, variable should be `poolPrice`, not `poolPirce`

Reply:
busted 😄 , i'll update that